### PR TITLE
Add a menu entry to show/hide hidden files in the filebrowser

### LINKED ISSUE #2049: Option to display hidden files in the file browser side panel
I couldn't find another issue that mentioned this, so I'll open it here:

It would be nice to be able to view hidden files in the file browser. Perhaps it could be in the `File` dropdown, although I don't see any other configuration options that it would group naturally with.  Maybe a standalone command in the `Commands` tab would be most appropriate.

### PR DESCRIPTION
**Pull request by loichuder**
The PR is still (very) rough around the edges but I draft it for discussion. Bear with my ignorance for my first dive in the JupyterLab depths :pray: 

## References

Fixes #2049 

## User-facing changes

A <s>button was added in the filebrowser toolbar</s> **entry in the View Menu** to toggle the visibility of hidden files:
Peek 2021-08-06 12-10

## Code changes

<s>The hidden files are shown/hidden using the CSS trick of @Alex-ley: https://github.com/jupyterlab/jupyterlab/issues/2049#issuecomment-852958811.</s>

To me, several points needs to be addressed before merging (don't hesitate to add more) but I would need some guidance :slightly_smiling_face: :
- [x] <s>First of all, how robust is the CSS trick using `li[data-is-dot]` ? Should it be done in another fashion ?</s> **Implemented in a new `FileBrowser` model.**
- [x] <s>Implement a **real** toggle button rather than using a simple button that do not give real visual feedback. Accessibility must also be addressed. Is there already a component in `@jupyterlab/ui-components` or `@lumino/widgets` that would do the job ?</s> **Implemented as an entry in ViewMenu instead as per [@jasongrout's  comment](https://github.com/jupyterlab/jupyterlab/pull/10769#issuecomment-894095526)**
- [x] <s>Having hidden files served requires JupyterLab to have been launched with `--ContentsManager.allow_hidden=true`. Should it now be the default ?</s> **The entry is disabled if `ContentsManager.allow_hidden` is not set to `true`.**
- [x] <s>I think it would also be handy for the visibility of hidden files to be persisted in a front-end setting.</s> **Added a `showHiddenFiles` setting in file-browser extension**.

## Backwards-incompatible changes

None that I know of.

### CODE REVIEW

**Comment by fcollonval** [comment]
@loichuder as discussed during the latest the dev meeting, it will be good to bring that information from the server to the frontend.

To do that, the usual way is to add an information to the page config dictionary on the backend. Then that dictionary can be access in the frontend.

Here is an example for passing information on the shut down menu entry: https://github.com/jupyterlab/jupyterlab/pull/5226/files

Let's bring pointer to the current code as that PR is old.

The information should be transferred in the page config in the backend at https://github.com/jupyterlab/jupyterlab/blob/21b3fdf053701368569736bce7661ad3925236fb/jupyterlab/labapp.py#L718-L724

Then you can read it in the frontend like this: https://github.com/jupyterlab/jupyterlab/blob/42f28b3e33d92321b625cf815cdbbc79baed9a00/packages/mainmenu-extension/src/index.ts#L157

> To give you some details on how it works, the page config dictionary is dumped in the `index.html` to be read by the frontend:

https://github.com/jupyterlab/jupyterlab_server/blob/a67e80c653d0a63ed519becd9d30fd0ac1b07fda/jupyterlab_server/templates/index.html#L14

File: packages/filebrowser/src/browser.ts
Comment (krassowski) [inline packages/filebrowser/src/browser.ts:110]: The file listing rendering is complex and for performance reasons reuses existing HTML nodes; this means that we cannot just toggle visibility on nodes that currently exist or we will have a repeat of https://github.com/jupyterlab/jupyterlab/issues/8860 (but for hidden files, with far worse consequences) each time we refresh the listing (enter a directory/clicking "refresh", etc). That issue was fixed in
https://github.com/jupyterlab/jupyterlab/pull/10100 and you will also find a way to persist setting in there.

In general there are two ways to approach the problem:
- do not render files starting with dot in the listing
- hide files starting with dot with CSS but keep them in the listing

The former may have a performance benefit on rendering when hidden files are not shown (less DOM nodes for the browser to handle), but will require refresh on toggle; I would encourage you to go this route.

If however you want to go the CSS route (or as a general advice for future) I would recommend implementing it by toggling a class - let's name it `jp-mod-dot-hidden` - on the listing (or `FileBrowser` node which contains the listing) and having a CSS rule like:

```csss
.jp-mod-dot-hidden li[data-is-dot] {
    display: none
}
```

This way you avoid unnecessary query and iteration on (potentially dozens or hundreds of) DOM nodes and let the browser handle it with the most optimized code for this purpose you can get. Still I recommend not showing these nodes in the first place.

**Review by krassowski — COMMENTED** [review]
>  First of all, how robust is the CSS trick using li[data-is-dot] ? Should it be done in another fashion ?

Yes it should be done differently to avoid bugs when changing directories or adding/removing files at a minimum, but for performance reasons I would even try to avoid CSS solutions and do not render these nodes in the first place. Please see my comment in the code review section.

>  Implement a real toggle button rather than using a simple button that do not give real visual feedback. Accessibility must also be addressed. Is there already a component in @jupyterlab/ui-components or @lumino/widgets that would do the job ?

That's right, thank you for thinking about accessibility. Please see the `pressed` state proposed in https://github.com/jupyterlab/jupyterlab/pull/10727 and let me know if this would work for you here. Do you have a specific set of icons in mind? While [file-hidden](https://materialdesignicons.com/icon/file-hidden) may look good I do not know if it has a pair so that it is colour blind friendly... There is also [eye-off](https://materialdesignicons.com/icon/eye-off] and it could be paired with [eye](https://materialdesignicons.com/icon/eye) (and there are outline versions too). What do other editors use as an indicator icon for hidden files?

**Comment by loichuder** [comment]
Thank you @fcollonval and @krassowski for the pointers, that is very helpful !

> Please see the `pressed` state proposed in #10727 and let me know if this would work for you here. 

This is great, I think that's exactly what is needed here. 

> Do you have a specific set of icons in mind? 

Not yet, thanks for the proposals :slightly_smiling_face: . [file-hidden](https://materialdesignicons.com/icon/file-hidden) and the pair [eye-off](https://materialdesignicons.com/icon/eye-off)/[eye](https://materialdesignicons.com/icon/eye) are indeed promising. [file-eye](https://materialdesignicons.com/icon/file-eye) could also do the trick. I will have to test them out.

**Comment by loichuder** [comment]
Addressed all issues except the button aspect. Waiting on https://github.com/jupyterlab/jupyterlab/pull/10727 for this.

**Comment by jasongrout** [comment]
Thanks for working on this!

Given that [most people don't find this a problem (from our survey)](https://github.com/jupyterlab/jupyterlab/issues/2049#issuecomment-788514357), I think a menu item (maybe in the View menu) would be more appropriate than a toolbar button. The file browser toolbar space is extremely limited, so I think there should be a very high bar to get on it.

> Not required IMO but I think it would also be handy for the visibility of hidden files to be persisted in a front-end setting.

I likewise think it should be a persistent setting in the frontend.

**Comment by jasongrout** [comment]
> I think a menu item (maybe in the View menu) would be more appropriate than a toolbar button.

Another place people might look for a setting like this is the context menu for the filebrowser, though that is getting kind of long these days too.

**Comment by loichuder** [comment]
> Addressed all issues except the button aspect. Waiting on #10727 for this.

Given that a `ViewMenu` entry was used, we don't need to discuss the icons nor the visual aspect of the button. Therefore, this is ready for review.

As the decision for the context menu entry was not definitive, I did not add it for now.

**Comment by jasongrout** [comment]
> Given that [most people don't find this a problem (from our survey)](https://github.com/jupyterlab/jupyterlab/issues/2049#issuecomment-788514357),

I want to thank @layne-sadler again for the work he did to summarize survey results and put the data on various issues. That was a lot of work, and we appreciate it!

**Comment by fcollonval** [comment]
@loichuder woud you mind rebasing this PR (to update to the new UI test structure) and correct the UI test (the view menu is changing)?

**Comment by loichuder** [comment]
@fcollonval I rebased and squashed my commits. 

Not sure how to fix the UI tests, though. Can you give me some pointers to run and fix the tests (https://jupyterlab.readthedocs.io/en/stable/developer/contributing.html#visual-regression-and-ui-tests is a little bare-bones :confused: ) ?

**Comment by loichuder** [comment]
Oh and would it make sense to set `ContentsManager.allow_hidden` to `True` in the JupyterLab default config ? 

The entry on the front-end will be unchecked which will keep the previous behaviour but it will be enabled so that people can change it ?

**Comment by fcollonval** [comment]
The UI test comparison looks like this (after is left and before is right):

image

That makes me wonder if we should not set `isVisible` on the command so that the menu entry is not there if the server setting is not set. Otherwise it may be confusing user that it is possible to display them somehow from the frontend.

@isabela-pf @jasongrout what do you think?

---

A note on how to update the screenshot for the UI tests:

```
cd galata
jlpm run start:detached
jlpm run test:update
```

If you don't want to run all tests, you can either pass the file test you want to run or use a selector with the `-g` option:

```
jlpm run test:update test/jupyterlab/menus.test.ts
jlpm run test:update -g "Open menu item View"
```

**Comment by jasongrout** [comment]
> That makes me wonder if we should not set `isVisible` on the command so that the menu entry is not there if the server setting is not set. Otherwise it may be confusing user that it is possible to display them somehow from the frontend.

 I think that sounds great.

**Comment by krassowski** [comment]
> That makes me wonder if we should not set `isVisible` on the command so that the menu entry is not there if the server setting is not set. Otherwise it may be confusing user that it is possible to display them somehow from the frontend.

Discoverability-wise keeping it visible but disabled sounds better to me.

**Comment by isabela-pf** [comment]
I think I agree with @krassowski that I'd prefer it disabled but visible in the menu. It can be helpful for people to know that is an option even if it is not currently configured.

My main questions are 
1. How would this server setting be configured so that the menu item is available? Is that configuration discoverable or is this essentially going to add a button that is disabled for most users because they don't know how to set it up?
2. Are there any consequences for users if they were to make hidden files visible by accident? That would be the only reason I could think of to put more in the way of accessing this feature, to ensure that people only select it when they truly intend to.

--
(Also thanks to @krassowski for reviewing for accessibility above! I'm extremely grateful.)

**Comment by fcollonval** [comment]
Some additional information (and personal opinion):
- features like shut down and terminals are related to server configuration. If the configuration disable them the UI elements will not be visible
- from my experience most users don't have a clue about how jupyterlab works (aka running a server). And they usually don't know about the server configuration - at best they know about some CLI options. So it will be complex for them to set the configuration.
- in managed environment (like JupyterHub), the user often can't customize the server config.

**Comment by isabela-pf** [comment]
@fcollonval that makes sense, and I've had similar experiences with a majority of users knowing less about the inner workings of JupyterLab (like I'm one to judge 😆). Especially since there are other cases where features relying on server configuration don't appear in the menus when disabled, I would be okay with the pattern continuing here. Thanks so much for the clarification!

**Comment by blink1073** [comment]
Hi @loichuder, are you available to finish up this PR?  We've created the 3.2.x and are headed toward a release

**Comment by loichuder** [comment]
> Hi @loichuder, are you available to finish up this PR? We've created the 3.2.x and are headed toward a release

Yes, thanks for the ping. 

To sum up discussions above, I should make the entry disappear instead of disabled if the server does not allow serving hidden files. 

However, as the default config of the JupyterLab server is to not serve hidden files, most users won't be able to benefit from this feature :disappointed:.

So, I would be in favour of setting `ContentsManager.allow_hidden` to `True` in the JupyterLab default config so that the button appear and that the display of hidden files is handled by the front-end setting. I know I mentioned this several times so sorry to insist but I did not get a clear approval :sweat_smile:

**Comment by blink1073** [comment]
I think it would make sense to include a setting for `ContentsManager.allow_hidden` in `jupyter_lab_config.d/jupyterlab.json` in the `jupyter-config` [folder](https://github.com/jupyterlab/jupyterlab/tree/05a2fa047c5416ae1dd6a6cdac165c13a3513f47/jupyter-config).  It would also have to be included in [`setup.py`](https://github.com/jupyterlab/jupyterlab/blob/05a2fa047c5416ae1dd6a6cdac165c13a3513f47/setup.py#L31).

**Comment by loichuder** [comment]
In a new file, you mean ? Should it not be set in `jupyter_server_config.d/jupyterlab.json` ?

**Comment by blink1073** [comment]
You're right, since `ContentsManager` is defined in the server, it makes sense to go in that file.

**Review by blink1073 — APPROVED** [review]
Thank you!

**Comment by loichuder** [comment]
Wait before merging, I don't think https://github.com/jupyterlab/jupyterlab/pull/10769/commits/87da0ed87d37c3354d13fa67476766053774ff06 did the trick. I think it was working due to a local leftover config on my end.

How is `jupyter-config/jupyter_server_config.d` supposed to be picked up by the app ? It does not appear in `jupyter --paths`.

**Comment by blink1073** [comment]
@loichuder if you `pip uninstall -y jupyterlab` until it says it is not installed, then `pip install .` (*not* editable mode), the `data_files` should get moved and you should see it in `<env>/etc/jupyter`

**Comment by loichuder** [comment]
I could not make it work by adding the config field to `jupyter-config/jupyter_server_config.d/jupyterlab.json`. 

After some [research](https://github.com/jupyter/enhancement-proposals/blob/master/28-jupyter-server/jupyter-server.md#configuration-system), the files in `jupyter_server_config.d` are apparently [only to list and enable extensions ?](https://github.com/Zsailer/jupyter_config_overview/tree/master/server-1.x/%7Bsys_prefix%7D/etc/jupyter/jupyter_server_config.d). Meaning it cannot change `jupyter_server` configuration ? Is that accurate/still true ?

I worked around this by adding a new `jupyter_server_config.json` to `jupyter-config` but it would be great to have feedback from someone with more insight on the configuration setup.

**Comment by blink1073** [comment]
Ah, that is right.  I think we would need to update the way we are loading application data to use config.d files.

**Comment by fcollonval** [comment]
Unfortunately we should revert ba2c756 as this may overwrite environment settings of the user. And I would suggest merging without that option turn on by default for now and open an issue for adding support for `config.d` file (unsure if the best place is `JupyterApp` in `jupyter_core` or `ServerApp` in `jupyter-server`).

**Comment by blink1073** [comment]
My thought would be `JupyterApp` should handle `config.d`

**Review by fcollonval — APPROVED** [review]
Thanks @loichuder LGTM

**Comment by blink1073** [comment]
@meeseeksdev please backport to 3.2.x

**Comment by ellisonbg** [comment]
Want to make sure, but it looks like this will be in 3.2?

**Comment by blink1073** [comment]
> Want to make sure, but it looks like this will be in 3.2?

That is correct

**Comment by fperez** [comment]
Just came here to say how grateful I am for the work on this!! While it may be a bit obscure for many users, it's also quite critical for a non-zero subset of them (a group I happen to belong to :), so I am delighted to see it go in. 🚀 !

**Pull request merged**
